Create SIP-22: Deep Links (#134)
MetaMask/SIPs: SIPS/sip-22.md

author: Christian Montoya (@Montoya)
created: 2024-03-20
---

## Abstract

This SIP describes a URL scheme for Snaps entry points. This scheme can be used by Snap or dapp methods to trigger the entry point and navigate the user to that location. 

## Motivation

While some Snap entry points are reactive, meaning they can be triggered by a method, others can only be triggered by a user navigating to the location where that entry point is displayed. In order for a Snap or dapp to programmatically navigate a user to that entry point, the Snap or dapp needs a way to identify that location. The location is identified by a URL scheme, which is called a _deep link_.

## Specification

### Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)

### URL Scheme 

A URL for a deep link uses the format: 

`metamask://[Snap ID]/page/[page name]`

Where: 

- `metamask` refers to the host application 
- `[Snap ID]` refers to the canonical ID of the installed Snap

> - `page` refers to a page inside of the application 

Reviewer: Seems unnecessary given that the deep link is always inside the application.

```suggestion
```

> - `[page name]` refers to the canonical name of an individual page 

Reviewer 2: Maybe add a list of pages that can be used.

### Method for Navigating to a Deep Link

In order for a Snap or dapp to navigate a user to a deep link, a method is provided: 

> `snap_open`. 

Reviewer 2: `snap_*` methods are only supported by Snaps. We may want to have a `wallet_*` method instead?

Author: Actually this makes me think the Snap should have the method internally and expose a custom RPC method to dapps if it wants to allow dapps to do this.

Author: I made the method specific to Snaps but I am not actually certain whether it would be best to make it a `snap_*` or `wallet_*` method. Would like to hear your perspective based on the latest draft.

Reviewer: My thoughts regarding this is if we want to expose this functionality to dapps, then we should make this a `wallet_*` method. Reason is because otherwise you end up having an extraneous request to the snap. It now has to be called for the invoke snap and then it's called for the `onHomePage` call.

This method takes one parameter, which is the URL of the deep link: 

```typescript
await snap.request({
  method: 'snap_open',
  params: {

>     url: 'metamask://npm:@consensys/starknet-snap/page/home'

Reviewer 2: This is not a valid URL. It will navigate to `metamask://consensys/starknet-snap/page/home`.

Author: Oh well, maybe we have to encode the Snap ID then. I was hoping not to do that but seems like we don't have much choice....

Reviewer: Might have to encode the snap ID, as I said in standup the url is still not good without `metamask://`, maybe we can ship a url builder function to make it easier for devs.

  },
});
```

### Examples

The proposed URL for the Starknet Snap home page: 

> `metamask://npm:@consensys/starknet-snap/page/home`

Author: Tough question. One possibility is allowing URLs like: 

`localhost:8000/page/home` 

Another is to say that any URL without the snap ID, like just `/page/home` will be interpreted as referring to the current Snap, and will work on both localhost and production. 

I like either approach, what do others think?

> `metamask://npm:@consensys/starknet-snap/page/home`

Reviewer: What do you mean by current snap?

### URL Scheme 

[URL syntax](https://en.wikipedia.org/wiki/URL#Syntax) is defined with a scheme, authority, path, and with optional (query and fragment) portions. This SIP defines a new scheme: `metamask:`, the syntax is defined below:

`metamask://[authority]/[path]`

Where: 

- `[authority]` refers to either `extension` or `snap`
- `[path]` refers to the entire path which depending on the authority can be different
  - For the `extension` authority, the following paths are available:
    - `/` - links to the extension's home page
  - For the `snap` authority, the path starts with the snap ID and has the following paths available to it:
    - `/home` - leads to the snap's [home page](/SIPS/sip-15.md) (which is its settings page if it doesn't have a home page)

**Note:** In the future, fragments can potentially be used for navigation to specific portions of a page and params can be provided from a calling method for exports such as `onHome`.

### Examples

The proposed URL for the Starknet Snap home page: 

`metamask://snap/npm:@consensys/starknet-snap/home`

The URL for navigating to the extension's home page:

> `metamask://extension/`

Author: This is good but I don't want to make it client specific. Why not: 

`metamask://client/`

That would work on mobile too.

Author: This would also align with the terminology we use elsewhere like `snap_getClientStatus`

Reviewer: Oh right, forgot we have something called mobile 😂

---
sip: 22
title: MetaMask URL scheme
status: Draft

> author: Christian Montoya (@Montoya), Hassan Malik(@hmalik88)

Reviewer 2: Nit:
```suggestion
author: Christian Montoya (@Montoya), Hassan Malik (@hmalik88)
```

## Specification

### Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)

### URL Scheme 

[URL syntax](https://en.wikipedia.org/wiki/URL#Syntax) is defined with a scheme, authority, path, and with optional (query and fragment) portions. This SIP defines a new scheme: `metamask:`, the syntax is defined below:

`metamask://[authority]/[path]`

Where: 

- `[authority]` refers to either `client` or `snap` (in the current case, `client` can mean either the extension or mobile version of MetaMask)
- `[path]` refers to the entire path which depending on the authority can be different
  - For the `client` authority, the following paths are available:
    - `/` - links to the client's home page
  - For the `snap` authority, the path starts with the snap ID and has the following paths available to it:
    - `/home` - leads to the snap's [home page](/SIPS/sip-15.md) (which is its settings page if it doesn't have a home page)

> **Note:** 
> 1. In the future, fragments can potentially be used for navigation to specific portions of a page and params can be provided from a calling method for exports such as `onHome`.
> 2. This URL scheme assumes parity between `client` locations.

Reviewer 2: Nit:
```suggestion
> [!NOTE]
> 1. In the future, fragments can potentially be used for navigation to specific portions of a page and params can be provided from a calling method for exports such as `onHome`.
> 2. This URL scheme assumes parity between `client` locations.
```

### Examples

The proposed URL for the Starknet Snap home page: 

`metamask://snap/npm:@consensys/starknet-snap/home`

> The URL for navigating to the extension's home page:

Reviewer 2: Nit:
```suggestion
The URL for navigating to the client's home page:
```

Reviewer: The language was intentional because it is actually an example url of how you would navigate to the extension's home page.

Reviewer 2: Not on mobile where this will eventually be supported.

`metamask://client/`

### Using Deep Links

> A Snap can currently use navigation from within the snaps ui `Link` component. In the future, a SIP will be drafted to allow for programmatic navigation with another RPC method.

Reviewer 2: We don't need to speculate about future SIPs here imo.
```suggestion
A Snap will be able to navigate to these links using the SIP-7 `Link` component.
```
